File: KrishiDukaan-V2/app/dashboard/_lib/category-info.ts
/**
 * Category-specific product information schema.
 *
 * Each category defines the structured fields shown in the create/edit form
 * and rendered on the product detail page.  The data is stored as:
 *   product.categoryInfo: Record<string, string | string[]>
 * where "chips" field types are stored as string[] and everything else as string.
 */

// ─── Category list ───────────────────────────────────────────────────────────

export const PRODUCT_CATEGORIES = [
  "Fertilizers",
  "Pesticides",
  "Herbicides",
  "Bio-Stimulants",
  "Seeds",
  "Sprayers",
  "Tools",
  "Other",
] as const;

export type ProductCategory = (typeof PRODUCT_CATEGORIES)[number];

/** Returns true if the value is one of the canonical categories. */
export function isStandardCategory(value: string): value is ProductCategory {
  return (PRODUCT_CATEGORIES as readonly string[]).includes(value);
}

// ─── Field schema ─────────────────────────────────────────────────────────────

export type FieldType = "text" | "textarea" | "chips";

export interface CategoryField {
  key: string;
  label: string;
  type: FieldType;
  placeholder?: string;
}

// ─── Per-category field definitions ──────────────────────────────────────────

export const CATEGORY_FIELDS: Record<ProductCategory, CategoryField[]> = {
  Fertilizers: [
    { key: "application", label: "Application",          type: "textarea", placeholder: "e.g. Suitable for foliar spray and fertigation..." },
    { key: "dosage",      label: "Recommended Dosage",   type: "text",     placeholder: "e.g. 3–5 gm / Litre" },
    { key: "bestForCrops",label: "Best For Crops",       type: "chips",    placeholder: "e.g. Tomatoes, Wheat, Sugarcane" },
  ],

  Pesticides: [
    { key: "activeIngredient", label: "Active Ingredient",    type: "text",  placeholder: "e.g. Chlorpyrifos 50% EC" },
    { key: "chemicalGroup",    label: "Chemical Group",       type: "text",  placeholder: "e.g. Organophosphate" },
    { key: "targetPest",       label: "Target Pest",          type: "text",  placeholder: "e.g. Aphids, Thrips, White Flies" },
    { key: "modeOfAction",     label: "Mode of Action",       type: "textarea", placeholder: "e.g. Contact and systemic action..." },
    { key: "dosage",           label: "Recommended Dosage",   type: "text",  placeholder: "e.g. 2 ml / Litre of water" },
    { key: "waitingPeriod",    label: "Waiting Period",       type: "text",  placeholder: "e.g. 14 days before harvest" },
    { key: "bestForCrops",     label: "Best For Crops",       type: "chips", placeholder: "e.g. Cotton, Rice, Vegetables" },
  ],

  Herbicides: [
    { key: "activeIngredient",   label: "Active Ingredient",             type: "text",  placeholder: "e.g. Glyphosate 41% SL" },
    { key: "targetWeeds",        label: "Target Weeds",                  type: "text",  placeholder: "e.g. Broad-leaf weeds, Grasses" },
    { key: "type",               label: "Type",                          type: "text",  placeholder: "Selective / Non-Selective" },
    { key: "applicationStage",   label: "Application Stage",             type: "text",  placeholder: "Pre-Emergence / Post-Emergence" },
    { key: "dosage",             label: "Recommended Dosage",            type: "text",  placeholder: "e.g. 1.5 L / acre" },
    { key: "bestForCrops",       label: "Best For Crops",                type: "chips", placeholder: "e.g. Wheat, Maize, Soybean" },
  ],

  "Bio-Stimulants": [
    { key: "keyIngredients",   label: "Key Ingredients",     type: "text",     placeholder: "e.g. Humic acid, Seaweed extract" },
    { key: "benefits",         label: "Benefits",            type: "chips",    placeholder: "e.g. Improves root development, Enhances nutrient uptake" },
    { key: "applicationMethod",label: "Application Method",  type: "text",     placeholder: "e.g. Soil drench, foliar spray" },
    { key: "dosage",           label: "Recommended Dosage",  type: "text",     placeholder: "e.g. 2–3 ml / Litre" },
    { key: "growthStage",      label: "Growth Stage",        type: "text",     placeholder: "e.g. Vegetative, Flowering" },
    { key: "bestForCrops",     label: "Best For Crops",      type: "chips",    placeholder: "e.g. Tomatoes, Grapes, Paddy" },
  ],

  Seeds: [
    { key: "varietyName",     label: "Variety Name",      type: "text", placeholder: "e.g. HYV-123, Pusa Basmati" },
    { key: "seedType",        label: "Seed Type",         type: "text", placeholder: "Hybrid / Open Pollinated / OPV" },
    { key: "germinationRate", label: "Germination Rate",  type: "text", placeholder: "e.g. 90–95%" },
    { key: "maturityPeriod",  label: "Maturity Period",   type: "text", placeholder: "e.g. 90–110 days" },
    { key: "seedRate",        label: "Seed Rate",         type: "text", placeholder: "e.g. 8–10 kg / acre" },
    { key: "suitableSeason",  label: "Suitable Season",   type: "text", placeholder: "Kharif / Rabi / Zaid" },
    { key: "bestRegions",     label: "Best Regions",      type: "chips",placeholder: "e.g. Maharashtra, Karnataka, Punjab" },
  ],

  Sprayers: [
    { key: "tankCapacity", label: "Tank Capacity",  type: "text", placeholder: "e.g. 16 Litres" },
    { key: "material",     label: "Material",       type: "text", placeholder: "e.g. HDPE, Stainless Steel" },
    { key: "weight",       label: "Weight",         type: "text", placeholder: "e.g. 2.5 kg (empty)" },
    { key: "powerSource",  label: "Power Source",   type: "text", placeholder: "Manual / Battery / Petrol" },
    { key: "sprayRange",   label: "Spray Range",    type: "text", placeholder: "e.g. 5–8 metres" },
    { key: "nozzleType",   label: "Nozzle Type",    type: "text", placeholder: "e.g. Adjustable fan nozzle" },
  ],

  Tools: [
    { key: "material",   label: "Material",      type: "text", placeholder: "e.g. High-carbon steel" },
    { key: "dimensions", label: "Dimensions",    type: "text", placeholder: "e.g. 40 × 15 cm" },
    { key: "weight",     label: "Weight",        type: "text", placeholder: "e.g. 1.2 kg" },
    { key: "suitableFor",label: "Suitable For",  type: "text", placeholder: "e.g. Weeding, Digging" },
    { key: "handleType", label: "Handle Type",   type: "text", placeholder: "e.g. Wooden, Fibreglass" },
    { key: "warranty",   label: "Warranty",      type: "text", placeholder: "e.g. 1 year manufacturer warranty" },
  ],

  Other: [
    { key: "keyFeatures",      label: "Key Features / Product Info", type: "textarea", placeholder: "e.g. Non-ionic surfactant. Improves spreading and sticking of spray solutions..." },
    { key: "benefits",         label: "Benefits",                    type: "chips",    placeholder: "e.g. Reduces surface tension, Improves coverage, Enhances absorption" },
    { key: "applicationMethod",label: "Application Method",          type: "text",     placeholder: "e.g. Tank mix with pesticide or herbicide solution" },
    { key: "dosage",           label: "Recommended Dosage",          type: "text",     placeholder: "e.g. 0.5 ml / Litre of spray solution" },
    { key: "specifications",   label: "Specifications",              type: "textarea", placeholder: "List key product specifications..." },
    { key: "application",      label: "Application",                 type: "textarea", placeholder: "Describe how the product is used..." },
    { key: "additionalNotes",  label: "Additional Notes",            type: "textarea", placeholder: "Any other relevant information..." },
  ],
};

// ─── Helpers ──────────────────────────────────────────────────────────────────

/** Chips fields that are stored as string[] instead of string. */
export const CHIPS_FIELDS = new Set(["bestForCrops", "bestRegions", "benefits"]);

// ─── Firestore-backed schema override ────────────────────────────────────────
//
// PRODUCT_CATEGORIES / CATEGORY_FIELDS above remain the compile-time source of
// truth for TYPES, and the fallback when Firestore is unreachable. At runtime
// the list is loaded from `settings/productSchema` — the same document the
// mobile app reads — so the two platforms can't drift apart, and a category can
// be added without shipping an app release.
//
// This is deliberately additive: nothing that imports PRODUCT_CATEGORIES today
// breaks, and callers opt in by using getProductCategories()/getCategoryFields().

let loadedCategories: string[] | null = null;
let loadedFields: Record<string, CategoryField[]> | null = null;
let loadedChipsFields: Set<string> | null = null;
let loadPromise: Promise<void> | null = null;

/**
 * Loads settings/productSchema once per session. Safe to call repeatedly —
 * concurrent callers share one in-flight request, and any failure silently
 * leaves the hardcoded defaults in place.
 */
export async function loadProductSchema(): Promise<void> {
  if (loadPromise) return loadPromise;
  loadPromise = (async () => {
    try {
      const { doc, getDoc } = await import("firebase/firestore");
      const { db } = await import("../../firebase");
      const snap = await getDoc(doc(db, "settings", "productSchema"));
      if (!snap.exists()) return;
      const data = snap.data() as Record<string, unknown>;

      const rawCategories = data.categories;
      if (!Array.isArray(rawCategories) || rawCategories.length === 0) return;

      const names: string[] = [];
      const fields: Record<string, CategoryField[]> = {};
      for (const entry of rawCategories) {
        const name = String((entry as any)?.name ?? "").trim();
        if (!name) continue;
        names.push(name);
        const rawFields = (entry as any)?.fields;
        fields[name] = Array.isArray(rawFields)
          ? rawFields
              .map((f: any) => ({
                key: String(f?.key ?? ""),
                label: String(f?.label ?? ""),
                type: (["text", "textarea", "chips"].includes(String(f?.type))
                  ? String(f.type)
                  : "text") as FieldType,
                placeholder: f?.placeholder ? String(f.placeholder) : undefined,
              }))
              .filter((f: CategoryField) => f.key.length > 0)
          : [];
      }
      if (names.length === 0) return;

      loadedCategories = names;
      loadedFields = fields;
      if (Array.isArray(data.chipsFields) && data.chipsFields.length > 0) {
        loadedChipsFields = new Set(data.chipsFields.map(String));
      }
    } catch {
      // Offline / rules / malformed doc — keep the hardcoded defaults.
    }
  })();
  return loadPromise;
}

/** Category list for dropdowns — Firestore's when loaded, else the constant. */
export function getProductCategories(): readonly string[] {
  return loadedCategories ?? PRODUCT_CATEGORIES;
}

/**
 * Category Info fields for a category name. Unlike CATEGORY_FIELDS this
 * accepts ANY string (including a category added on mobile or via Firestore
 * that this build has no type for) and falls back to the "Other" field set,
 * which is what the form already does for a custom category.
 */
export function getCategoryFields(category: string): CategoryField[] {
  if (loadedFields) {
    return loadedFields[category] ?? loadedFields["Other"] ?? CATEGORY_FIELDS.Other;
  }
  return isStandardCategory(category)
    ? CATEGORY_FIELDS[category]
    : CATEGORY_FIELDS.Other;
}

/** Chips-typed field keys — Firestore's when loaded, else the constant. */
export function getChipsFields(): Set<string> {
  return loadedChipsFields ?? CHIPS_FIELDS;
}

/** Convert a raw Firestore record to a typed CategoryInfo map. */
export function parseCategoryInfo(
  raw: Record<string, unknown>,
): Record<string, string | string[]> {
  const out: Record<string, string | string[]> = {};
  for (const [k, v] of Object.entries(raw)) {
    if (Array.isArray(v)) out[k] = v.map(String);
    else if (v != null) out[k] = String(v);
  }
  return out;
}

/**
 * Backward-compatibility helper: synthesize a categoryInfo map from the legacy
 * flat fertilizer fields that were stored directly on product documents
 * before this refactor (nitrogen, phosphorus, potassium, applicationDesc,
 * dosage, bestForCrops).
 *
 * Returns the synthesized map, or null if no legacy fields are present.
 */
export function synthesizeFertilizerInfo(
  data: Record<string, unknown>,
): Record<string, string | string[]> | null {
  const n  = data.nitrogen   ? String(data.nitrogen)   : "";
  const p  = data.phosphorus ? String(data.phosphorus) : "";
  const k  = data.potassium  ? String(data.potassium)  : "";
  const app= data.applicationDesc ? String(data.applicationDesc) : "";
  const dos= data.dosage     ? String(data.dosage)     : "";
  const bfc= Array.isArray(data.bestForCrops)
    ? data.bestForCrops.map(String)
    : data.bestForCrops ? [String(data.bestForCrops)] : [];

  if (!n && !p && !k && !app && !dos && !bfc.length) return null;

  const out: Record<string, string | string[]> = {};
  if (n)         out.nitrogen    = n;
  if (p)         out.phosphorus  = p;
  if (k)         out.potassium   = k;
  if (app)       out.application = app;
  if (dos)       out.dosage      = dos;
  if (bfc.length)out.bestForCrops= bfc;
  return out;
}

/**
 * Get the effective categoryInfo for a product, falling back to legacy flat
 * fields for Fertilizer products that predate this refactor.
 */
export function effectiveCategoryInfo(
  data: Record<string, unknown>,
): Record<string, string | string[]> | null {
  if (data.categoryInfo && typeof data.categoryInfo === "object" &&
      !Array.isArray(data.categoryInfo)) {
    const ci = data.categoryInfo as Record<string, unknown>;
    if (Object.keys(ci).length > 0) return parseCategoryInfo(ci);
  }
  // Fall back to legacy flat fields for Fertilizers
  const cat = String(data.category ?? "").trim();
  if (!cat || cat === "Fertilizers" || cat.toLowerCase().includes("fertilizer")) {
    return synthesizeFertilizerInfo(data);
  }
  return null;
}

File: KrishiDukaan-V2/mobile/lib/core/utils/category_info.dart
/// Category-specific product information schema — Dart port of
/// `app/dashboard/_lib/category-info.ts` so the mobile "Product Insights"
/// section shows the same per-category spec fields as the web product page.
library;

import '../models/catalog_model.dart';

const List<String> productCategories = [
  'Fertilizers',
  'Pesticides',
  'Herbicides',
  'Bio-Stimulants',
  'Seeds',
  'Sprayers',
  'Tools',
  'Other',
];

bool isStandardCategory(String value) => productCategories.contains(value);

enum CategoryFieldType { text, textarea, chips }

class CategoryField {
  final String key;
  final String label;
  final CategoryFieldType type;
  const CategoryField(this.key, this.label, this.type);
}

const Map<String, List<CategoryField>> categoryFields = {
  'Fertilizers': [
    CategoryField('application', 'Application', CategoryFieldType.textarea),
    CategoryField('dosage', 'Recommended Dosage', CategoryFieldType.text),
    CategoryField('bestForCrops', 'Best For Crops', CategoryFieldType.chips),
  ],
  'Pesticides': [
    CategoryField('activeIngredient', 'Active Ingredient', CategoryFieldType.text),
    CategoryField('chemicalGroup', 'Chemical Group', CategoryFieldType.text),
    CategoryField('targetPest', 'Target Pest', CategoryFieldType.text),
    CategoryField('modeOfAction', 'Mode of Action', CategoryFieldType.textarea),
    CategoryField('dosage', 'Recommended Dosage', CategoryFieldType.text),
    CategoryField('waitingPeriod', 'Waiting Period', CategoryFieldType.text),
    CategoryField('bestForCrops', 'Best For Crops', CategoryFieldType.chips),
  ],
  'Herbicides': [
    CategoryField('activeIngredient', 'Active Ingredient', CategoryFieldType.text),
    CategoryField('targetWeeds', 'Target Weeds', CategoryFieldType.text),
    CategoryField('type', 'Type', CategoryFieldType.text),
    CategoryField('applicationStage', 'Application Stage', CategoryFieldType.text),
    CategoryField('dosage', 'Recommended Dosage', CategoryFieldType.text),
    CategoryField('bestForCrops', 'Best For Crops', CategoryFieldType.chips),
  ],
  'Bio-Stimulants': [
    CategoryField('keyIngredients', 'Key Ingredients', CategoryFieldType.text),
    CategoryField('benefits', 'Benefits', CategoryFieldType.chips),
    CategoryField('applicationMethod', 'Application Method', CategoryFieldType.text),
    CategoryField('dosage', 'Recommended Dosage', CategoryFieldType.text),
    CategoryField('growthStage', 'Growth Stage', CategoryFieldType.text),
    CategoryField('bestForCrops', 'Best For Crops', CategoryFieldType.chips),
  ],
  'Seeds': [
    CategoryField('varietyName', 'Variety Name', CategoryFieldType.text),
    CategoryField('seedType', 'Seed Type', CategoryFieldType.text),
    CategoryField('germinationRate', 'Germination Rate', CategoryFieldType.text),
    CategoryField('maturityPeriod', 'Maturity Period', CategoryFieldType.text),
    CategoryField('seedRate', 'Seed Rate', CategoryFieldType.text),
    CategoryField('suitableSeason', 'Suitable Season', CategoryFieldType.text),
    CategoryField('bestRegions', 'Best Regions', CategoryFieldType.chips),
  ],
  'Sprayers': [
    CategoryField('tankCapacity', 'Tank Capacity', CategoryFieldType.text),
    CategoryField('material', 'Material', CategoryFieldType.text),
    CategoryField('weight', 'Weight', CategoryFieldType.text),
    CategoryField('powerSource', 'Power Source', CategoryFieldType.text),
    CategoryField('sprayRange', 'Spray Range', CategoryFieldType.text),
    CategoryField('nozzleType', 'Nozzle Type', CategoryFieldType.text),
  ],
  'Tools': [
    CategoryField('material', 'Material', CategoryFieldType.text),
    CategoryField('dimensions', 'Dimensions', CategoryFieldType.text),
    CategoryField('weight', 'Weight', CategoryFieldType.text),
    CategoryField('suitableFor', 'Suitable For', CategoryFieldType.text),
    CategoryField('handleType', 'Handle Type', CategoryFieldType.text),
    CategoryField('warranty', 'Warranty', CategoryFieldType.text),
  ],
  'Other': [
    CategoryField('keyFeatures', 'Key Features / Product Info', CategoryFieldType.textarea),
    CategoryField('benefits', 'Benefits', CategoryFieldType.chips),
    CategoryField('applicationMethod', 'Application Method', CategoryFieldType.text),
    CategoryField('dosage', 'Recommended Dosage', CategoryFieldType.text),
    CategoryField('specifications', 'Specifications', CategoryFieldType.textarea),
    CategoryField('application', 'Application', CategoryFieldType.textarea),
    CategoryField('additionalNotes', 'Additional Notes', CategoryFieldType.textarea),
  ],
};

/// Fields stored as a list of strings instead of a plain string.
const Set<String> chipsFields = {'bestForCrops', 'bestRegions', 'benefits'};

/// Synthesizes a categoryInfo-shaped map from the legacy flat fertilizer
/// fields that predate the categoryInfo refactor (nitrogen, phosphorus,
/// potassium, applicationDesc, dosage, bestForCrops). Returns null if none
/// of those fields are present.
Map<String, dynamic>? synthesizeFertilizerInfo(CatalogModel catalog) {
  final n = catalog.nitrogen;
  final p = catalog.phosphorus;
  final k = catalog.potassium;
  final app = catalog.applicationDesc?.trim() ?? '';
  final dos = catalog.dosage?.trim() ?? '';
  final bfc = catalog.bestForCrops ?? const <String>[];

  if (n == null && p == null && k == null && app.isEmpty && dos.isEmpty && bfc.isEmpty) {
    return null;
  }

  return {
    if (n != null) 'nitrogen': n.toString(),
    if (p != null) 'phosphorus': p.toString(),
    if (k != null) 'potassium': k.toString(),
    if (app.isNotEmpty) 'application': app,
    if (dos.isNotEmpty) 'dosage': dos,
    if (bfc.isNotEmpty) 'bestForCrops': bfc,
  };
}

/// Effective categoryInfo for a product: the doc's own `categoryInfo` map if
/// set, else a synthesized map from legacy flat Fertilizer fields, else null.
Map<String, dynamic>? effectiveCategoryInfo(CatalogModel catalog) {
  if (catalog.categoryInfo != null && catalog.categoryInfo!.isNotEmpty) {
    return catalog.categoryInfo;
  }
  final cat = catalog.category.trim();
  if (cat.isEmpty || cat == 'Fertilizers' || cat.toLowerCase().contains('fertilizer')) {
    return synthesizeFertilizerInfo(catalog);
  }
  return null;
}
